content/blog/callpackage-a-tool-for-the-lazy.md

```diff
@@ -0,0 +1,219 @@
+nicely greet you.
+
+As you can see, `writeShellScriptBin` gets passed in by `callPackage` automatically.
+
+For this simple setup, having to create an extra file, seems to be a lot of
+boilerplate, though if you continue reading, you will see, it is worth it!
+
+## 1. Benefit: parametrized builds
+
+Now lets change the `default.nix`:
+
+```nix
+# default.nix
+let pkgs = import <nixpkgs> { }; in
+{
+  hello = pkgs.callPackage ./hello.nix { };
+}
+```
+
+Now we build using `nix-build -A hello`, the outcome will be the same as above.
+
+Now to "parametrize" the build we also change the `hello.nix` a bit:
+
+```nix
+# hello.nix
+{ writeShellScriptBin
+, audience ? "world"
+}:
+writeShellScriptBin "hello" ''
+  echo "hello, ${audience}!"
+''
+```
+
+Building this will still yield the same output as before, though now things get
+interesting, alter your `default.nix` yet another time:
+
+```nix
+# default.nix
+let pkgs = import <nixpkgs> { }; in
+{
+  hello = pkgs.callPackage ./hello.nix { };
+  people = pkgs.callPackage ./hello.nix { audience = "people"; };
+}
+```
+
+Building via `nix-build -A people` will now yield a script that prints "hello,
+people" instead.
+
+We can use the very same syntax to also overwrite the automatically discovered
+arguments like `writeShellScriptBin`, though that doesn't make sense here.
+
+Though, for a Go program that expects `buildGoModule` it is common to see some
+expression like `callPackage ./go-program.nix { buildGoModule = buildGo116Module; }`
+to enforce a certain Go compiler version.
+
+## 2. Benefit: overrides
+
+As a consequence from the parametrized builds, we can also change the value of
+the parameters after the fact, using the derivations `override` function.
+
+Consider this new `default.nix`:
+
+```nix
+# default.nix
+let pkgs = import <nixpkgs> { }; in
+rec {
+  hello = pkgs.callPackage ./hello.nix { };
+  people = pkgs.callPackage ./hello.nix { audience = "people"; };
+  folks = hello.override { audience = "folks"; };
+}
+```
+
+Building and running the `folks` attribute, will give again a new version of the
+script.
+
+All the other parameters will remain the same as the have been when `hello` was
+instantiated.
+
+This is especially useful and often seen on packages that provide a whole lot of
+options to optimize the build.
+
+An example to mention here is the `neovim` attribute in nixpkgs, which has has 
+some overrideable arguments like `extraLuaPackages`, `extraPythonPackages`, or
+`withRuby`.
+
+## 3. Benefit: modifiable
+
+And now I want to introduce one of my most favorite benefits:
+
+You can actually create your own version of `callPackage`, which comes in quite
+handy when you have large sets where the attributes to be built depend on each
+other.
+
+```nix
+# In the next examples I will not implement or show the "called" 
+# files, as I think they are not necessary to understand the point I
+# want to make.
+```
```

Comment:
As above


```suggestion
```nix
> **Note**
> In the following examples I will not implement or show the "called packages", as I think they are not necessary to understand the point I want to make.
```
```

```suggestion
```nix
# In the following examples I will not implement or show the "called 
# packages", as I think they are not necessary to understand the point I
# want to make.
```
```

Reply:
Changed in 1d65f8b381e3e1ff397288ddd6b0a3b653857f5a